text/2978-stack_vec.md

```diff
@@ -0,0 +1,185 @@
+
+Instead of relying on a heap-allocator, stack-based memory area is added and removed on-demand in a last-in-first-out (LIFO) order according to the calling workflow of a program. Let's illustrate an imaginary function `super()` that allocates 32 bits before calling another function named `sub()` that also allocates 32 bits of auxiliary memory:
+
+```txt
+0 bits           32 bits          64 bits   
+
+
+|       |   ->   |       |   ->   |       |
+---------        ---------        ---------
+                 | super |        |  sub  |
+                 ---------        ---------
+                                  | super |
+                                  ---------
+```
+
+* Now `super()` returns and the exactly same thing happens in reverse order, dropping everything when out of scope.
+
+```txt
+64 bits          32 bits          0 bits   
+
+
+|       |   ->   |       |   ->   |       |
+---------        ---------        ---------
+|  sub  |        | super |
+---------        ---------
+| super |
+---------
+```
+
+`StackVec` takes advantage of this predictable behavior to reserve an exactly amount of uninitialized bytes up-front and these bytes form a buffer where elements can be included dynamically.
+
+```rust
+fn main() {
+    // `stack_vec` has a pre-allocated memory of 2048 bits that can store up to 64 decimals.
+    let mut stack_vec: StackVec<i32, 64> = StackVec::new();
+
+    // Although reserved, there isn't anything explicitly stored yet
+    assert_eq!(stack_vec.len(), 0);
+
+    // Initializes the first 32 bits with a simple '1' decimal or
+    // 00000000 00000000 00000000 00000001 bits
+    stack_vec.push(1);
+
+    // Our vector memory is now split into a 32/2016 pair of initialized and
+    // uninitialized memory respectively
+    assert_eq!(stack_vec.len(), 1);
+}
+```
+
+Of course, fixed buffers lead to some inflexibility because unlike `Vec`, the underlying capacity can not expand at run-time and there will never be more than 64 elements in the above example.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The most natural module for `StackVec` is `core::collections` with an API that basically mimics most of the current `Vec<T>` surface.
+
+```rust
+// core::collections
+
+pub struct StackVec<T, const N: usize> {
+    data: MaybeUninit<[T, N]>,
+    len: usize
+}
+
+impl<T, const N: usize> StackVec<T, N> {
+    // Much of the `Vec` API goes here
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### Additional complexity
+
+New and existing users are likely to find it difficult to differentiate the purpose of each vector type, especially people that don't have a theoretical background in memory management.
+
+### The current ecosystem is fine
+
+Even with all the fragmentation, different types of memory usage is an overkill in certain situations. If someone wants to use stack memory in an embedded application, then it is just a matter of grabbing an appropriated crate.
```

Comment:
Note that fragmentation is only really relevant when the type in question frequently is part of the API of other libraries. If it isn't, at worst you have a bit of code bloat when multiple similar crates are used by your dependencies.